src/lib/form-field/form-field.ts

```diff
@@ -0,0 +1,382 @@
+
+
+/** An interface which allows a control to work inside of a `MdFormField`. */
+export abstract class MdFormFieldControl {
+  /**
+   * Stream that emits whenever the state of the control changes such that the parent `MdFormField`
+   * needs to run change detection.
+   */
+  stateChanges: Observable<void>;
+
+  /** Gets the element ID for this control. */
+  abstract getId(): string;
+
+  /** Gets the placeholder for this control. */
+  abstract getPlaceholder(): string;
+
+  /** Gets the NgControl for this control. */
+  abstract getNgControl(): NgControl | null;
+
+  /** Whether the control is focused. */
+  abstract isFocused(): boolean;
+
+  /** Whether the control is empty. */
+  abstract isEmpty(): boolean;
+
+  /** Whether the control is required. */
+  abstract isRequired(): boolean;
+
+  /** Whether the control is disabled. */
+  abstract isDisabled(): boolean;
+
+  /** Whether the control is in an error state. */
+  abstract isErrorState(): boolean;
+
+  /** Sets the list of element IDs that currently describe this control. */
+  abstract setDescribedByIds(ids: string[]): void;
+
+  /** Focuses this control. */
+  abstract focus(): void;
+}
+
+
+/** Container for form controls that applies Material Design styling and behavior. */
+@Component({
+  moduleId: module.id,
+  selector: 'md-input-container, mat-input-container, md-form-field, mat-form-field',
+  templateUrl: 'form-field.html',
+  // MdInput is a directive and can't have styles, so we need to include its styles here.
+  // The MdInput styles are fairly minimal so it shouldn't be a big deal for people who aren't using
+  // MdInput.
+  styleUrls: ['form-field.css', '../input/input.css'],
+  animations: [
+    trigger('transitionMessages', [
+      state('enter', style({ opacity: 1, transform: 'translateY(0%)' })),
+      transition('void => enter', [
+        style({ opacity: 0, transform: 'translateY(-100%)' }),
+        animate('300ms cubic-bezier(0.55, 0, 0.55, 0.2)'),
+      ]),
+    ]),
+  ],
+  host: {
+    // Remove align attribute to prevent it from interfering with layout.
+    '[attr.align]': 'null',
+    'class': 'mat-input-container mat-form-field',
+    '[class.mat-input-invalid]': '_control.isErrorState()',
+    '[class.mat-form-field-invalid]': '_control.isErrorState()',
+    '[class.mat-focused]': '_control.isFocused()',
+    '[class.ng-untouched]': '_shouldForward("untouched")',
+    '[class.ng-touched]': '_shouldForward("touched")',
+    '[class.ng-pristine]': '_shouldForward("pristine")',
+    '[class.ng-dirty]': '_shouldForward("dirty")',
+    '[class.ng-valid]': '_shouldForward("valid")',
+    '[class.ng-invalid]': '_shouldForward("invalid")',
+    '[class.ng-pending]': '_shouldForward("pending")',
+    '(click)': '_control.focus()',
+  },
+  encapsulation: ViewEncapsulation.None,
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+
+export class MdFormField implements AfterViewInit, AfterContentInit, AfterContentChecked {
```

Comment:
In `public_api.ts`, we should re-export this as `MdInputContainer` to stay backwards compatible with anyone importing it (about 18 instances within Google)